godot-ffi/src/godot_ffi.rs

```diff
@@ -487,3 +487,67 @@ mod scalars {
         }
     }
 }
+
+unsafe impl<T> GodotFfi for Option<T>
+where
+    T: GodotFfi,
+{
+    fn sys(&self) -> sys::GDExtensionTypePtr {
+        match self {
+            Some(value) => value.sys(),
+            None => {
+                let nil_ptr = ptr::null_mut();
+                let new_nil = interface_fn!(variant_new_nil);
+
+                unsafe {
+                    new_nil(nil_ptr);
+                }
+
+                nil_ptr as sys::GDExtensionTypePtr
+            }
+        }
+    }
+
+    unsafe fn from_sys(ptr: sys::GDExtensionTypePtr) -> Self {
+        let nil_ptr = ptr::null_mut();
+
+        interface_fn!(variant_new_nil)(nil_ptr);
+
+        if ptr as sys::GDExtensionVariantPtr == nil_ptr {
+            return None;
+        }
+
+        Some(T::from_sys(ptr))
+    }
+
+    unsafe fn from_sys_init(init_fn: impl FnOnce(sys::GDExtensionTypePtr)) -> Self {
+        let mut raw = std::mem::MaybeUninit::uninit();
+        init_fn(raw.as_mut_ptr() as sys::GDExtensionTypePtr);
+
+        Self::from_sys(raw.assume_init())
```

Comment:
I have no experience with this, but I was assuming that `raw` is either a variant nil or T and passing it on to `from_sys` will handle both these cases.

Reply:
the MaybeUninit is basically allocating space for the init function to initialize a new value of that type. since it's here a pointer we're only allocating enough space for a pointer, but lots of types are bigger than a pointer so i dont think it'll work for them.

what happens if we try to write say a Vector3 into the MaybeUninit?

Reply:
I get your point. I think we can disregard this once we ensure that T can only be a pointer. After looking into Godot types again, it appears that only object types are nullable. Everything else is not nullable.

Reply:
given the right safety bounds on a marker-trait, this should be correct then. it is at least equivalent to a null-checked version of what `ffi_methods` does for a type where the pointer is used as the opaque value.